File: BOJ/Gold V/BOJ5557.py
#BOJ5557 1학년 20210512
import sys
input = sys.stdin.readline

def main():
    n = int(input())
    li = list(map(int, input().rstrip().split()))
    dp = [[0 for _ in range(21)] for _ in range(n-1)]
    dp[0][li[0]] = 1
    for i in range(n-2):
        nextNum = li[i+1]
        for j in range(21):
            if j + nextNum <= 20: dp[i+1][j+nextNum] += dp[i][j]
            if j - nextNum >= 0: dp[i+1][j-nextNum] += dp[i][j]
    print(dp[-1][li[-1]])

if __name__ == '__main__':
    main()

File: BOJ/Gold V/BOJ5557.java
//BOJ5557 1학년 20210709
import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

import static java.lang.Math.*;

public class Main {
    private static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    private static int n;
    private static int[] arr;
    private static long[][] dp;

    public static void main(String[] args) throws IOException {
        n = Integer.parseInt(br.readLine());
        arr = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

        dp = new long[n-1][21];
        dp[0][arr[0]] = 1;
        for(int i = 1; i < n-1; i++){
            int v = arr[i];
            for(int j = 0; j <= 20; j++){
                if(dp[i-1][j] > 0){
                    if(j + v <= 20) dp[i][j + v] += dp[i-1][j];
                    if(j - v >= 0) dp[i][j - v] += dp[i-1][j];
                }
            }
        }
        System.out.println(dp[n-2][arr[n-1]]);
    }

}